packages/periphery/contracts/oracles/PendleOracle.sol

```diff
@@ -0,0 +1,143 @@
+    address pendleMarket,
+    address secondaryPoolOracle,
+    address yieldToken,
+    uint16 secondsAgo,
+    uint16 secondsAgoLiquidation,
+    bool usePtToSyOracle
+  ) external onlyOwner {
+    if (secondsAgo == 0 || secondsAgoLiquidation == 0) revert WrongValue();
+    if (
+      quoteToken == address(0) ||
+      baseToken == address(0) ||
+      pendleMarket == address(0) ||
+      secondaryPoolOracle == address(0) ||
+      yieldToken == address(0)
+    ) {
+      revert ZeroAddress();
+    }
+
+    _assertOracleIsInitialized(pendleMarket, secondsAgo);
+    _assertOracleIsInitialized(pendleMarket, secondsAgoLiquidation);
+
+    (IStandardizedYield sy, , ) = PendleMarketV3(pendleMarket).readTokens();
+    uint8 ptDecimals = IERC20Metadata(baseToken).decimals();
+    uint8 syDecimals = IERC20Metadata(address(sy)).decimals();
+
+    getParams[quoteToken][baseToken] = OracleParams({
+      pendleMarket: pendleMarket,
+      secondsAgo: secondsAgo,
+      secondsAgoLiquidation: secondsAgoLiquidation,
+      yieldToken: yieldToken,
+      secondaryPoolOracle: secondaryPoolOracle,
+      ptDecimals: ptDecimals,
+      syDecimals: syDecimals,
+      usePtToSyOracle: usePtToSyOracle
+    });
+  }
+
+  /// @notice check Pendle oracle is initialized - https://docs.pendle.finance/Developers/Integration/HowToIntegratePtAndLpOracle#third-initialize-the-oracle
+  function _assertOracleIsInitialized(address pendleMarket, uint16 secondsAgo) private {
+    (bool increaseCardinalityRequired, , bool oldestObservationSatisfied) = pendle.getOracleState(
+      pendleMarket,
+      secondsAgo
+    );
+    if (increaseCardinalityRequired) revert PendlePtLpOracleIsNotInitialized(secondsAgo);
+    if (!oldestObservationSatisfied) revert PendlePtLpOracleIsNotInitialized(secondsAgo);
+  }
+
+  function getBalancePrice(address quoteToken, address baseToken) external view returns (uint256) {
+    return _getPriceX96(quoteToken, baseToken, false);
+  }
+
+  function getMargincallPrice(address quoteToken, address baseToken) external view returns (uint256) {
+    return _getPriceX96(quoteToken, baseToken, true);
+  }
+
+  function _getPriceX96(
+    address quoteToken,
+    address baseToken,
+    bool isMargincallPrice
+  ) private view returns (uint256 priceX96) {
+    OracleParams storage poolParams = getParams[quoteToken][baseToken];
+    if (poolParams.pendleMarket == address(0)) revert ZeroAddress();
+
+    // PT - base token
+    // SY - wrapped yield token
+    // YQT - yield quote token (e.g. ezETH)
+    // QT - quote token (e.g. WETH)
+
+    // Pendle market: PT / SY
+    // 1.0 SY == 1.0 YQT
+    // secondary pool: YQT / QT
+
+    // getPtToSyRate() and getPtToAssetRate() returns price with 18 decimals
+    uint256 pendlePrice = 0;
+
+    if (poolParams.usePtToSyOracle) {
+      pendlePrice = pendle.getPtToSyRate(
+        poolParams.pendleMarket,
+        isMargincallPrice ? poolParams.secondsAgoLiquidation : poolParams.secondsAgo
+      );
+    } else {
```

Comment:
I think there is some mistake here: the next step of multiplying this price by the one of `ibToken/quoteToken` pair is definitely incorrect. Not sure if you need to return this value in x96 format or to use some other extra oracle in the next step

Reply:
I think the best way is to remove this `if (poolParams.usePtToSyOracle)` branching completely and assume we always want the `pt/sy` price here. We may sync on this matter later